src/parser.c

```diff
@@ -0,0 +1,168 @@
+/*                                                                            */
+/* ************************************************************************** */
+
+/**
+ * @file parser.c
+ * @brief 
+ */
+
+#include "mod_parser.h"
+
+/**
+ * @brief 			Driver function to parse token list.
+ * 
+ * Creates cmd structures and fills them with the token list.
+ * @param lst 		Pointer to list of tokens.
+ * @param cmd 		Pointer pointer to first node of cmd list.
+ * @return t_err 	ERR_MALLOC, SUCCESS
+ */
+t_err	ft_parser(t_tkn_list *lst, t_cmd **cmd)
+{
+	t_err	err;
+	t_cmd	*new;
+	bool	exe_found;
+	bool	cmd_complete;
+
+	exe_found = 0;
+	cmd_complete = 0;
+	new = NULL;
+	err = ft_init_cmd(&new);
+	if (err != SUCCESS)
+		return (err);
+	while (lst)
+	{
+		err = ft_categorise(&lst, new, &exe_found, &cmd_complete);
+		if (err != SUCCESS)
+			return (err);
+		if (cmd_complete)
+			new = ft_new_cmd(new, cmd, &exe_found, &cmd_complete);
+		if (!new)
+			return (ERR_MALLOC);
+		lst = lst->next;
+	}
+	return (ft_finish_cmd_list(new, cmd));
+}
+
+/**
+ * @brief Adds current cmd to cmd_list and creates a new cmd struct.
+ * 
+ * @param curr 			Current cmd struct.
+ * @param cmd 			List of cmd structs.
+ * @param exe_found 	Bool to indicate if executable has been found.
+ * @param cmd_complete 	Bool to indicate if cmd ist complete.
+ * @return t_cmd* 
+ */
+t_cmd	*ft_new_cmd(t_cmd *curr, t_cmd **cmd, bool *exe_found,
+	bool *cmd_complete)
+{
+	t_cmd	*new;
+	t_err	err;
+
+	new = NULL;
+	curr->args = ft_split(curr->arg_buf, ' ');
+	curr->delims = ft_split(curr->delim_buf, ' ');
+	if (!curr->args || !curr->delims)
+		return (NULL);
+	ft_add_cmd(curr, cmd);
+	err = ft_init_cmd(&new);
+	if (err != SUCCESS)
+		return (NULL);
+	*exe_found = 0;
+	*cmd_complete = 0;
+	return (new);
+}
+
+/**
+ * @brief Initiates the cmd struct.
+ * 
+ * @param new Newly created cmd struct.
+ * @return t_err 	ERR_MALLOC, SUCCESS
+ */
+t_err	ft_init_cmd(t_cmd **new)
+{
+	t_cmd	*tmp;
+
+	tmp = malloc(sizeof(t_cmd));
+	if (!tmp)
+		return (ERR_MALLOC);
+	tmp->append = 0;
+	tmp->arg_buf = NULL;
+	tmp->args = NULL;
+	tmp->delim_buf = NULL;
+	tmp->delims = NULL;
+	tmp->exe = NULL;
+	tmp->fd_in = -1;
+	tmp->fd_out = -1;
+	tmp->infile = 0;
+	tmp->outfile = 0;
+	tmp->index = -1;
+	tmp->next = NULL;
+	*new = tmp;
+	return (SUCCESS);
+}
+
+/**
+ * @brief Sorts token and calls appropriate functions
+ * 			to save token into the cmd structure.
+ * 
+ * @param lst 			Pointer to current token in the list.
+ * @param new 			Pointer to cmd structure.
+ * @param exe_found 	Bool to indicate if executable has been found.
+ * @param cmd_complete 	Bool to indicate if cmd ist complete.
+ * @return t_err 		SUCCESS
+ */
+t_err	ft_categorise(t_tkn_list **lst, t_cmd *new, bool *exe_found,
+	bool *cmd_complete)
+{
+	t_tkn_list	*tmp;
+	t_err		err;
+
+	err = SUCCESS;
+	tmp = *lst;
+	if (ft_strncmp(tmp->content, "<", ft_strlen(tmp->content)) == 0)
```

Comment:
Wait a minute, don't we expand now before lexing? Why should there be an empty string inside of a token?

Reply:
You mean parsing? :D 
Yeah and thats why an empty string can occur: "hello $mum" gets lexed into two tokens 
-  hello
-  $mum

If $mum is not set it still gets expanded resulting in an empty string (which is bash behavior). So the two token would be 
- hello 
- "" (empty string)

Now you miss my ft_strcmp :0

Reply:
Oh! Or did you mean lexing (= which is tokenize?) 
Right now we lex->expand. I still need to implement field splitting which would not affect the problem: since an empty string can't be split it would remain as such.
We could implement a cleaner function which would search for empty strings and delete them from the tkn_lst.

Reply:
Yeah I mean lexing, so have everything expanded before tokenising, so that the actual input gets saved in the token list. Is there any argument against that order?

Reply:
I wanted to implement it as next step so the sequence would be
Lexing - Expand (excl quote removal) - Field splitting (if Expand happened) - Quote Removal 

If Lexing is before expand it's easier for here-doc delims (they can be skipped). 
But this could also be an extra check in the expand step.

But more importantly: echo needs the empty strings. If you echo "hello $VAR world" and VAR is not set echo prints "hello  world" (two spaces) since it prints a space after each arg if there is another one.